Type inference for union of arrays broken

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  Version 3.1.0-dev.20180925

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** array this context type compatibility assignability

**Code**

```ts
interface Array<T> {
    equalsShallow<T>(this: ReadonlyArray<T>, other: ReadonlyArray<T>): boolean;
}

type ExtractArrayType<TArray> = TArray extends Array<infer T> ? T : never;

[
    { a: ["x"], b: [] },
    { a: [], b: ["x"] },
    { a: ["x", 2], b: ["x", 2] },
    { a: ["1"], b: [1] },
    { a: [1], b: ["1"] },
    { a: [null], b: [undefined] },
    { a: [undefined], b: [null] },
    { a: [{}], b: [{}] },
].forEach(({ a, b }) => {

    // Compiles:
    const assignedA: typeof a = b;
    const assignedB: typeof b = a;

    // Compiles (prove that ReadonlyArray<>-types are compatible):
    const assignedAReadonly: ReadonlyArray<ExtractArrayType<typeof a>> = b;
    const assignedBReadonly: ReadonlyArray<ExtractArrayType<typeof b>> = a;

    // error TS2684: The 'this' context of type '(string | number)[] | null[] | undefined[] | {}[]'
    // is not assignable to method's 'this' of type 'ReadonlyArray<string | number>':
    a.equalsShallow(b);
});
```

Works in TS 3.0.

Note that the error disappears when changing `equalsShallow()` to `equalsShallow<T>(this: Array<T>, other: Array<T>): boolean`.

Further note that the error disappears when extracting a type for `a` and `b` (instead of using the literal type inference).

**Expected behavior:** No compile error

**Actual behavior:** Compile error, see snippet

**Playground Link:** [Link](https://www.typescriptlang.org/play/#src=interface%20Array%3CT%3E%20%7B%0D%0A%20%20%20%20equalsShallow%3CT%3E(this%3A%20ReadonlyArray%3CT%3E%2C%20other%3A%20ReadonlyArray%3CT%3E)%3A%20boolean%3B%0D%0A%7D%0D%0A%0D%0Atype%20ExtractArrayType%3CTArray%3E%20%3D%20TArray%20extends%20Array%3Cinfer%20T%3E%20%3F%20T%20%3A%20never%3B%0D%0A%0D%0A%5B%0D%0A%20%20%20%20%7B%20a%3A%20%5B%22x%22%5D%2C%20b%3A%20%5B%5D%20%7D%2C%0D%0A%20%20%20%20%7B%20a%3A%20%5B%5D%2C%20b%3A%20%5B%22x%22%5D%20%7D%2C%0D%0A%20%20%20%20%7B%20a%3A%20%5B%22x%22%2C%202%5D%2C%20b%3A%20%5B%22x%22%2C%202%5D%20%7D%2C%0D%0A%20%20%20%20%7B%20a%3A%20%5B%221%22%5D%2C%20b%3A%20%5B1%5D%20%7D%2C%0D%0A%20%20%20%20%7B%20a%3A%20%5B1%5D%2C%20b%3A%20%5B%221%22%5D%20%7D%2C%0D%0A%20%20%20%20%7B%20a%3A%20%5Bnull%5D%2C%20b%3A%20%5Bundefined%5D%20%7D%2C%0D%0A%20%20%20%20%7B%20a%3A%20%5Bundefined%5D%2C%20b%3A%20%5Bnull%5D%20%7D%2C%0D%0A%20%20%20%20%7B%20a%3A%20%5B%7B%7D%5D%2C%20b%3A%20%5B%7B%7D%5D%20%7D%2C%0D%0A%5D.forEach((%7B%20a%2C%20b%20%7D)%20%3D%3E%20%7B%0D%0A%0D%0A%20%20%20%20%2F%2F%20Works%3A%0D%0A%20%20%20%20const%20assignedA%3A%20typeof%20a%20%3D%20b%3B%0D%0A%20%20%20%20const%20assignedB%3A%20typeof%20b%20%3D%20a%3B%0D%0A%0D%0A%20%20%20%20%2F%2F%20Prove%20that%20ReadonlyArray%3C%3E-types%20are%20compatible%3A%0D%0A%20%20%20%20const%20assignedAReadonly%3A%20ReadonlyArray%3CExtractArrayType%3Ctypeof%20a%3E%3E%20%3D%20b%3B%0D%0A%20%20%20%20const%20assignedBReadonly%3A%20ReadonlyArray%3CExtractArrayType%3Ctypeof%20b%3E%3E%20%3D%20a%3B%0D%0A%0D%0A%20%20%20%20%2F%2F%20error%20TS2684%3A%20The%20'this'%20context%20of%20type%20'(string%20%7C%20number)%5B%5D%20%7C%20null%5B%5D%20%7C%20undefined%5B%5D%20%7C%20%7B%7D%5B%5D'%0D%0A%20%20%20%20%2F%2F%20is%20not%20assignable%20to%20method's%20'this'%20of%20type%20'ReadonlyArray%3Cstring%20%7C%20number%3E'%3A%0D%0A%20%20%20%20a.equalsShallow(b)%3B%0D%0A%7D)%3B%0D%0A) (no error because it's TS 3.0)

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -1,3 +1,6 @@
+
+
+
 /* @internal */
 namespace ts {
     const ambientModuleSymbolRegex = /^".+"$/;
@@ -13460,6 +13463,7 @@ namespace ts {
             let symbolStack: Symbol[];
             let visited: Map<boolean>;
             let contravariant = false;
+            let bivariant = false;
             let propagationType: Type;
             let allowComplexConstraintInference = true;
             inferFromTypes(originalSource, originalTarget);
@@ -13546,7 +13550,9 @@ namespace ts {
                             }
                             if (priority === inference.priority) {
                                 const candidate = propagationType || source;
-                                if (contravariant) {
+                                // We make contravariant inferences only if we are in a pure contravariant position,
+                                // i.e. only if we have not descended into a bivariant position.
+                                if (contravariant && !bivariant) {
                                     inference.contraCandidates = append(inference.contraCandidates, candidate);
                                 }
                                 else {
@@ -13798,7 +13804,12 @@ namespace ts {
 
             function inferFromSignature(source: Signature, target: Signature, skipParameters: boolean) {
                 if (!skipParameters) {
+                    const saveBivariant = bivariant;
+                    const kind = target.declaration ? target.declaration.kind : SyntaxKind.Unknown;
+                    // Once we descend into a bivariant signature we remain bivariant for all nested inferences
+                    bivariant = bivariant || kind === SyntaxKind.MethodDeclaration || kind === SyntaxKind.MethodSignature || kind === SyntaxKind.Constructor;
                     forEachMatchingParameterType(source, target, inferFromContravariantTypes);
+                    bivariant = saveBivariant;
                 }
                 const sourceTypePredicate = getTypePredicateOfSignature(source);
                 const targetTypePredicate = getTypePredicateOfSignature(target);

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -1,6 +1,3 @@
-
-
-
 /* @internal */
 namespace ts {
     const ambientModuleSymbolRegex = /^".+"$/;
@@ -13553,10 +13550,10 @@ namespace ts {
                                 // We make contravariant inferences only if we are in a pure contravariant position,
                                 // i.e. only if we have not descended into a bivariant position.
                                 if (contravariant && !bivariant) {
-                                    inference.contraCandidates = append(inference.contraCandidates, candidate);
+                                    inference.contraCandidates = appendIfUnique(inference.contraCandidates, candidate);
                                 }
                                 else {
-                                    inference.candidates = append(inference.candidates, candidate);
+                                    inference.candidates = appendIfUnique(inference.candidates, candidate);
                                 }
                             }
                             if (!(priority & InferencePriority.ReturnType) && target.flags & TypeFlags.TypeParameter && !isTypeParameterAtTopLevel(originalTarget, <TypeParameter>target)) {

--- a/tests/baselines/reference/bivariantInferences.js
+++ b/tests/baselines/reference/bivariantInferences.js
@@ -1,0 +1,17 @@
+//// [bivariantInferences.ts]
+// Repro from #27337
+
+interface Array<T> {
+    equalsShallow<T>(this: ReadonlyArray<T>, other: ReadonlyArray<T>): boolean;
+}
+
+declare const a: (string | number)[] | null[] | undefined[] | {}[];
+declare const b: (string | number)[] | null[] | undefined[] | {}[];
+
+let x = a.equalsShallow(b);
+
+
+//// [bivariantInferences.js]
+"use strict";
+// Repro from #27337
+var x = a.equalsShallow(b);

--- a/tests/baselines/reference/bivariantInferences.symbols
+++ b/tests/baselines/reference/bivariantInferences.symbols
@@ -1,0 +1,31 @@
+=== tests/cases/conformance/types/typeRelationships/typeInference/bivariantInferences.ts ===
+// Repro from #27337
+
+interface Array<T> {
+>Array : Symbol(Array, Decl(lib.es5.d.ts, --, --), Decl(lib.es5.d.ts, --, --), Decl(bivariantInferences.ts, 0, 0))
+>T : Symbol(T, Decl(lib.es5.d.ts, --, --), Decl(bivariantInferences.ts, 2, 16))
+
+    equalsShallow<T>(this: ReadonlyArray<T>, other: ReadonlyArray<T>): boolean;
+>equalsShallow : Symbol(Array.equalsShallow, Decl(bivariantInferences.ts, 2, 20))
+>T : Symbol(T, Decl(bivariantInferences.ts, 3, 18))
+>this : Symbol(this, Decl(bivariantInferences.ts, 3, 21))
+>ReadonlyArray : Symbol(ReadonlyArray, Decl(lib.es5.d.ts, --, --))
+>T : Symbol(T, Decl(bivariantInferences.ts, 3, 18))
+>other : Symbol(other, Decl(bivariantInferences.ts, 3, 44))
+>ReadonlyArray : Symbol(ReadonlyArray, Decl(lib.es5.d.ts, --, --))
+>T : Symbol(T, Decl(bivariantInferences.ts, 3, 18))
+}
+
+declare const a: (string | number)[] | null[] | undefined[] | {}[];
+>a : Symbol(a, Decl(bivariantInferences.ts, 6, 13))
+
+declare const b: (string | number)[] | null[] | undefined[] | {}[];
+>b : Symbol(b, Decl(bivariantInferences.ts, 7, 13))
+
+let x = a.equalsShallow(b);
+>x : Symbol(x, Decl(bivariantInferences.ts, 9, 3))
+>a.equalsShallow : Symbol(equalsShallow, Decl(bivariantInferences.ts, 2, 20), Decl(bivariantInferences.ts, 2, 20), Decl(bivariantInferences.ts, 2, 20), Decl(bivariantInferences.ts, 2, 20))
+>a : Symbol(a, Decl(bivariantInferences.ts, 6, 13))
+>equalsShallow : Symbol(equalsShallow, Decl(bivariantInferences.ts, 2, 20), Decl(bivariantInferences.ts, 2, 20), Decl(bivariantInferences.ts, 2, 20), Decl(bivariantInferences.ts, 2, 20))
+>b : Symbol(b, Decl(bivariantInferences.ts, 7, 13))
+

--- a/tests/baselines/reference/bivariantInferences.types
+++ b/tests/baselines/reference/bivariantInferences.types
@@ -1,0 +1,26 @@
+=== tests/cases/conformance/types/typeRelationships/typeInference/bivariantInferences.ts ===
+// Repro from #27337
+
+interface Array<T> {
+    equalsShallow<T>(this: ReadonlyArray<T>, other: ReadonlyArray<T>): boolean;
+>equalsShallow : <T>(this: ReadonlyArray<T>, other: ReadonlyArray<T>) => boolean
+>this : ReadonlyArray<T>
+>other : ReadonlyArray<T>
+}
+
+declare const a: (string | number)[] | null[] | undefined[] | {}[];
+>a : (string | number)[] | null[] | undefined[] | {}[]
+>null : null
+
+declare const b: (string | number)[] | null[] | undefined[] | {}[];
+>b : (string | number)[] | null[] | undefined[] | {}[]
+>null : null
+
+let x = a.equalsShallow(b);
+>x : boolean
+>a.equalsShallow(b) : boolean
+>a.equalsShallow : (<T>(this: ReadonlyArray<T>, other: ReadonlyArray<T>) => boolean) | (<T>(this: ReadonlyArray<T>, other: ReadonlyArray<T>) => boolean) | (<T>(this: ReadonlyArray<T>, other: ReadonlyArray<T>) => boolean) | (<T>(this: ReadonlyArray<T>, other: ReadonlyArray<T>) => boolean)
+>a : (string | number)[] | null[] | undefined[] | {}[]
+>equalsShallow : (<T>(this: ReadonlyArray<T>, other: ReadonlyArray<T>) => boolean) | (<T>(this: ReadonlyArray<T>, other: ReadonlyArray<T>) => boolean) | (<T>(this: ReadonlyArray<T>, other: ReadonlyArray<T>) => boolean) | (<T>(this: ReadonlyArray<T>, other: ReadonlyArray<T>) => boolean)
+>b : (string | number)[] | null[] | undefined[] | {}[]
+

--- a/tests/cases/conformance/types/typeRelationships/typeInference/bivariantInferences.ts
+++ b/tests/cases/conformance/types/typeRelationships/typeInference/bivariantInferences.ts
@@ -1,0 +1,12 @@
+// @strict: true
+
+// Repro from #27337
+
+interface Array<T> {
+    equalsShallow<T>(this: ReadonlyArray<T>, other: ReadonlyArray<T>): boolean;
+}
+
+declare const a: (string | number)[] | null[] | undefined[] | {}[];
+declare const b: (string | number)[] | null[] | undefined[] | {}[];
+
+let x = a.equalsShallow(b);

